src/conv/heuristics/ai_heuristics.cpp

```diff
@@ -0,0 +1,445 @@
+            return boost::any_cast<uint64_t>(id);
+        });
+        if(miopen::IsLogging(LoggingLevel::Info2))
+        {
+            std::stringstream ss;
+            for(auto& id : db_sol)
+                ss << solver::Id{id}.ToString() << " ID:" << id << ", ";
+            MIOPEN_LOG_I2("Cached solvers: " << ss.str());
+        }
+        return db_sol;
+    }
+
+    MIOPEN_LOG_I2("Evaluating Heuristic");
+
+    std::vector<float> res = model->Forward(problem);
+    std::vector<std::pair<int, float>> sort_res(res.size());
+    // sorts result based upon magnitude of result in vector, returned from Model,
+    // paired with original index (idx). Sort magnitudes in descending order.
+    // Greater magnitude = better solver. Indexes (idx), which will be used to map to solvers,
+    // with greater corresponding magnitude are at front of the vector so they get priority.
+    for(auto idx = 0; idx < res.size(); idx++)
+        sort_res[idx] = {idx, res[idx]};
+    const auto cmp = [](const std::pair<int, float>& a, const std::pair<int, float>& b) -> bool {
+        return a.second > b.second;
+    };
+    std::sort(sort_res.begin(), sort_res.end(), cmp);
+
+    // map idx to solver id and then anysolver
+    std::vector<uint64_t> sol;
+    std::vector<boost::any> any_sol;
+    for(const auto& kinder : sort_res)
+    {
+        const auto id     = kinder.first;
+        const auto sol_id = solver::Id{model->metadata.solver_map.at(id)};
+        if(!sol_id.IsValid())
+        {
+            MIOPEN_LOG_I2("Invalid solver " << model->metadata.solver_map.at(id) << " removed");
+            continue;
+        }
+        sol.push_back(sol_id.Value());
+        any_sol.push_back(sol_id.Value());
+    }
+    db.StoreRecord(problem.conv_problem, any_sol);
+    if(miopen::IsLogging(LoggingLevel::Info2))
+    {
+        std::stringstream ss;
+        for(auto& id : sol)
+            ss << solver::Id{id}.ToString() << " ID:" << id << ", ";
+        MIOPEN_LOG_I2("Heuristic Result: " << ss.str());
+    }
+    return sol;
+}
+} // namespace immed_mode
+#endif // MIOPEN_ENABLE_AI_IMMED_MODE_FALLBACK
+
+#if MIOPEN_ENABLE_AI_KERNEL_TUNING
+namespace tuning {
+
+Metadata::Metadata(const std::string& arch, const std::string& solver)
+{
+    const nlohmann::json metadata =
+        common::LoadJSON(GetSystemDbPath() + "/" + arch + "_" + solver + "_metadata.ktn.model");
+    num_tuning_params = metadata["num_tuning_params"].get<std::size_t>();
+    tuning_decodings = metadata["decodings"]["tunings"].get<std::unordered_map<std::string, int>>();
+}
+
+class Model
+{
+public:
+    Metadata metadata;
+    Model(const std::string& arch, const std::string& solver)
+        : metadata(Metadata(arch, solver)),
+          encoder(fdeep::load_model(EncoderPath(arch, solver), true, fdeep::dev_null_logger)),
+          decoder(fdeep::load_model(DecoderPath(arch, solver), true, fdeep::dev_null_logger))
```

Comment:
`EncoderPath`  and `DecoderPath` should check for the existence of the file before returning and `MIOPEN_THROW` if it does not.